File: leetCode/LinkedList/remove-nth-node-from-end-of-list.js
/**
 * Definition for singly-linked list.
 * function ListNode(val, next) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.next = (next===undefined ? null : next)
 * }
 */
/**
 * @param {ListNode} head
 * @param {number} n
 * @return {ListNode}
 */
 var removeNthFromEnd = function (head, n) {
    let preNode = new ListNode(-Infinity);
    preNode.next = head;
    let result = preNode;
    let count = 0;
    let tail = head;
  
    while (count < n) {
      count++;
      tail = tail.next;
    }
  
    let removed = head;
    let prev = preNode;
  
    while (tail) {
      tail = tail.next;
      removed = removed.next;
      prev = prev.next;
    }
  
    prev.next = removed.next;
    result = result.next
    return result;
  };

File: leetCode/LinkedList/remove-nth-node-from-end-of-list.py
# 19. Remove Nth Node From End of List

# Given the head of a linked list, remove the nth node from the end of the list and return its head.

# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next

class Solution:
    def removeNthFromEnd(self, head: Optional[ListNode], n: int) -> Optional[ListNode]:
        dummyNode = ListNode(None,head)
        slow = dummyNode
        fast = head
        
        while(n > 0 and fast):
            fast = fast.next
            n-=1
            
        while(fast):
            slow = slow.next
            fast = fast.next
            
        slow.next = slow.next.next
        
        return dummyNode.next
            